test/hotspot/jtreg/runtime/os/TestTracePageSizes.java

```diff
@@ -0,0 +1,279 @@
+ * @requires os.family == "linux"
+ * @requires vm.gc != "Z"
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:-SegmentedCodeCache TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:-SegmentedCodeCache -XX:+UseLargePages TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:-SegmentedCodeCache -XX:+UseTransparentHugePages TestTracePageSizes
+ */
+
+/*
+ * @test id=with-G1
+ * @summary Run tests with G1
+ * @requires os.family == "linux"
+ * @requires vm.gc.G1
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseG1GC TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseG1GC -XX:+UseLargePages TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseG1GC -XX:+UseTransparentHugePages TestTracePageSizes
+*/
+
+/*
+ * @test id=with-Parallel
+ * @summary Run tests with Parallel
+ * @requires os.family == "linux"
+ * @requires vm.gc.Parallel
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseParallelGC TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseParallelGC -XX:+UseLargePages TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseParallelGC -XX:+UseTransparentHugePages TestTracePageSizes
+*/
+
+/*
+ * @test id=with-Serial
+ * @summary Run tests with Serial
+ * @requires os.family == "linux"
+ * @requires vm.gc.Serial
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseSerialGC TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseSerialGC -XX:+UseLargePages TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseSerialGC -XX:+UseTransparentHugePages TestTracePageSizes
+*/
+
+import java.io.File;
+import java.util.LinkedList;
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+// Check that page sizes logged match what is recorded in /proc/self/smaps.
+// For transparent huge pages the matching is best effort since we can't
+// know for sure what the underlying page size is.
+public class TestTracePageSizes {
+    // Store address ranges with known page size.
+    private static LinkedList<RangeWithPageSize> ranges = new LinkedList<>();
+    private static boolean debug;
+
+    // Parse /proc/self/smaps using a regexp capturing the address
+    // ranges, what page size they have and if they might use
+    // transparent huge pages. The pattern is not greedy and will
+    // match as little as possible so each "segment" in the file
+    // will generate a match.
+    private static void parseSmaps() throws Exception {
+        Pattern smapsPattern = Pattern.compile(RangeWithPageSize.smapsPatternString, Pattern.DOTALL);
+        Scanner smapsScanner = new Scanner(new File("/proc/self/smaps"));
```

Comment:
Would that work if mappings are concurrently modified? (I guess it has to otherwise one could not read reliably from /proc at all)

Reply:
I can't say for sure, not sure exactly how the file is consumed by the Scanner. The segments and the fields in those segments that we really care about should not change and therefore I think we should be fine.

Reply:
I looked at https://www.kernel.org/doc/Documentation/filesystems/proc.txt
```
Note: reading /proc/PID/maps or /proc/PID/smaps is inherently racy (consistent
output can be achieved only in the single read call).
This typically manifests when doing partial reads of these files while the
memory map is being modified.  Despite the races, we do provide the following
guarantees:

1) The mapped addresses never go backwards, which implies no two
   regions will ever overlap.
2) If there is something at a given vaddr during the entirety of the
   life of the smaps/maps walk, there will be some output for it.
```

So, if Scanner reads the file in one go its okay. 

I'd say lets see how this goes then. If this test fails intermittently, we know where to look.

Reply:
I did a very basic test in `jshell` and created a `Scanner` on a file that I then updated between calls to `nextLine()`. The output from the scanner did not include my additions, so at least for a small file is seems like the contents is read in one go. But that doesn't say that is always the case. 

Bu I agree, we know where to look if we see intermittent failures.